apps/omg_child_chain/lib/omg_child_chain/gas_price/history/server.ex

```diff
@@ -0,0 +1,160 @@
+# Copyright 2020 OmiseGO Pte Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+defmodule OMG.ChildChain.GasPrice.History.Server do
+  @moduledoc """
+  Responsible for creating the gas price history store and managing its records,
+  including fetching, transforming, inserting and pruning the gas price records.
+
+  Also provides subscription that triggers on history changes.
+
+  This module utilizes [`ets`](https://elixir-lang.org/getting-started/mix-otp/ets.html)
+  instead of GenServer for history storage for the following reasons:
+
+    1. Multiple pricing strategies will be accessing the history. ets allows those strategies to
+       access the records concurrently.
+
+    2. Fetching the history is RPC-intensive and unnecessary between GenServer crashes.
+       Using ets allows the records to persist across crashes.
+  """
+  use GenServer
+  require Logger
+
+  alias OMG.ChildChain.GasPrice.History
+  alias OMG.ChildChain.GasPrice.History.Fetcher
+  alias OMG.Eth.Encoding
+  alias OMG.Eth.EthereumHeight
+
+  @history_table :gas_price_history
+
+  @type state() :: %__MODULE__{
+          num_blocks: non_neg_integer(),
+          earliest_stored_height: non_neg_integer(),
+          latest_stored_height: non_neg_integer(),
+          history_ets: :ets.tid(),
+          subscribers: [pid()]
+        }
+
+  defstruct num_blocks: 200,
+            earliest_stored_height: 0,
+            latest_stored_height: 0,
+            history_ets: nil,
+            subscribers: []
+
+  @doc false
+  @spec all() :: History.t()
+  def all() do
+    # No need to go through the GenSever to fetch from ets.
+    :ets.tab2list(@history_table)
+  end
+
+  #
+  # GenServer initialization
+  #
+
+  @doc false
+  @impl GenServer
+  def init(opts) do
+    num_blocks = Keyword.fetch!(opts, :num_blocks)
+    event_bus = Keyword.fetch!(opts, :event_bus)
+    :ok = event_bus.subscribe({:root_chain, "ethereum_new_height"}, link: true)
+
+    # The ets table is not initialized with `:read_concurrency` because we are expecting interleaving
+    # reads and writes. See http://erlang.org/doc/man/ets.html
+    history_ets = :ets.new(@history_table, [:ordered_set, :protected, :named_table])
```

Comment:
>  isolate the owning process of ets, so the ets could still survive when the business logic handling process failed

Hmm I did not realize that this process will take owner ship of this ets, and couldn't find any mention in the docs? Do you have a link to something I can read? If not I'll try experiment it through iex.

> However, testing wise, it might still be very convenient to be able to inject a ets table.

Yeah you are right on this. Will update.

Reply:
probably this shows the ownership of the ets table: https://dev.to/strech/where-is-my-ets-table-3ff1 (from my random googling :p)

Here is the previous [video](https://www.infoq.com/presentations/Let-It-Crash/) and [slide](https://qconlondon.com/london-2011/qconlondon.com/dl/qcon-london-2011/slides/SteveVinoski_LetItCrashExceptWhenYouShouldnt.pdf) I mentioned above.

Reply:
And you are right!

```elixir
iex(5)> task = Task.async(fn ->
...(5)>   table = :ets.new(:some_history_table, [:ordered_set, :protected, :named_table])
...(5)>   IO.inspect("#{inspect(self())}: ets table created: #{inspect(table)}")
...(5)>
...(5)>   record = {:foo, "bar"}
...(5)>   true = :ets.insert(table, record)
...(5)>   IO.inspect("#{inspect(self())}: inserted #{inspect(record)} to #{inspect(table)}")
...(5)> end)
"#PID<0.128.0>: ets table created: :some_history_table"
"#PID<0.128.0>: inserted {:foo, \"bar\"} to :some_history_table"
%Task{
  owner: #PID<0.105.0>,
  pid: #PID<0.128.0>,
  ref: #Reference<0.575725472.2248409092.246859>
}

iex(6)>
nil

iex(7)> _ = Task.await(task)
"#PID<0.128.0>: inserted {:foo, \"bar\"} to :some_history_table"

iex(8)>
nil

iex(9)> IO.inspect("#{inspect(self())}: The async task is: #{inspect(Process.info(task.pid))}")
"#PID<0.105.0>: The async task is: nil"
"#PID<0.105.0>: The async task is: nil"

iex(10)> IO.inspect("#{inspect(self())}: :some_history_table has #{inspect(:ets.tab2list(:some_history_table))}")
** (ArgumentError) argument error
    (stdlib 3.12) :ets.match_object(:some_history_table, :_)
    (stdlib 3.12) ets.erl:763: :ets.tab2list/1
```